### PR TITLE
Import

### PR DESCRIPTION
**Pull request by lukasschreiber**

### CODE REVIEW

File: js/src/main/scala/webapp/pages/HomePage.scala [hunk 1]
```diff
@@ -131,6 +135,34 @@ case class HomePage() extends Page {
...
+              fileList(0)
+                .text()
+                .toFuture
+                .onComplete(value => {
+                  if (value.isFailure) {
+                    value.failed.get.printStackTrace()
+                    toaster.make(value.failed.get.getMessage.nn, ToastMode.Short, ToastType.Error)
+                  }
+                  val json = value.getOrElse("");
+                  importIndexedDBJson(json)(using repositories)
```
Comment (lukasschreiber) [inline js/src/main/scala/webapp/pages/HomePage.scala:162]: some kind of feedback is needed to show the user that the import was successful

File: shared/src/main/scala/webapp/repo/Repository.scala [hunk 2]
```diff
@@ -35,6 +39,36 @@ case class Repository[A](name: String, defaultValue: A)(using
     codec: JsonValueCodec[A],
 ) {
 
+  given mapCodec: JsonValueCodec[mutable.Map[String, A]] = JsonCodecMaker.make
+
+  given magicCodec: JsonValueCodec[Repository[A]] =
```
Comment (lukasschreiber) [inline shared/src/main/scala/webapp/repo/Repository.scala:44]: the codec should have a name that implies, that the codec should not be used for anything other than importing from json

File: js/src/main/scala/webapp/pages/HomePage.scala [hunk 3]
```diff
@@ -131,6 +135,34 @@ case class HomePage() extends Page {
...
+            val json = exportIndexedDBJson
+            downloadJson(s"reform-export-${(new js.Date()).toISOString()}.json", json)
+          }),
+        ),
+        input(tpe := "file", cls := "file-input w-full max-w-xs", idAttr := "import-file"),
+        button(
+          cls := "btn btn-active p-2 h-fit min-h-10 border-0",
+          "Import DB",
+          onClick.foreach(_ => {
+            val fileList = document.querySelector("#import-file").asInstanceOf[HTMLInputElement].files
```
Comment (Oshgnacknak) [inline js/src/main/scala/webapp/pages/HomePage.scala:151]: Im sure we don't need to query select thanks to `outwatch`. Just Let the user input into a `Var` - or does this not work with file selects?

File: utils.js [hunk 4]
```diff
@@ -15,3 +15,13 @@ export const usesTurn = async (connection) => {
 		stats.get(selectedLocalCandidate)?.candidateType === "relay"
 	);
 };
+
+export const downloadJson = (name, text) => {
```
Comment (Oshgnacknak) [inline utils.js:19]: we might want to try https://stackoverflow.com/a/7208039.
Seems cleaner to me

**Review by Oshgnacknak — CHANGES_REQUESTED** [review]

File: js/src/main/scala/webapp/pages/HomePage.scala (hunk 3 above)
Comment (lukasschreiber) [reply js/src/main/scala/webapp/pages/HomePage.scala:151]: Sure, should work

File: utils.js (hunk 4 above)
Comment (lukasschreiber) [reply utils.js:19]: Will try, I think I have tried a bunch of solutions for the PDF download and the one I used was the one that worked best

Comment (lukasschreiber) [reply utils.js:19]: Did not work for me, connection to the ws has been lost

File: js/src/main/scala/webapp/pages/HomePage.scala (hunk 3 above)
Comment (lukasschreiber) [reply js/src/main/scala/webapp/pages/HomePage.scala:151]: Does not work with a Var

**Pull request merged**